proposals/nnnn-multiple-transports.md

```diff
@@ -0,0 +1,262 @@
+    String appID, String autoActivateID, boolean callbackToUIThread, boolean preRegister, String sHashID,
+    List<BaseTransportConfig> transportConfigs) throws SdlException {
+    ....
+}
+```
+`SdlProxyBase` class will be tweaked to keep the list of transport configs, and create `SdlSession` instance that uses SdlConnectionSelector.
+
+Note that `BTTransportConfig` and `MultiplexTransportConfig` should not be specified at the same time, as both of the transport will try to use Bluetooth device. `SdlProxyBase` class should detect this error during construction.
+
+`TCPTransport` class is likely to be updated to support head unit discovery feature over Wi-Fi, which will be discussed in separate proposal.
+
+### Extension of Core
+
+First of all, the ID value returned by `ApplicationImpl::app_id()` has to be connection independent. Currently the value is generated from a pair of `connection_handle` and `session_id` (please refer to `ApplicationManagerImpl::RegisterApplication()` implementation). We propose to replace it with the value of App Instance ID and make it connection independent.
+
+Likewise, `connection_key` value in `application_manager::Message` class should be replaced with App Instance ID. Any implementation above ApplicationManager that uses `connection_key` as an identification of an app should be updated to use App Instance ID instead.
+
+Secondary, SDL Core introduces a new component called `ConnectionSelector`. It is placed in between `ConnectionHandler` and `ApplicationManager`. It implements following features:
+* Conversion between `connection_key` and App Instance ID
+  - It receives OnServiceStartedCallback and OnServiceEndedCallback from ConnectionHandler, and keeps relationship between App Instance ID and `connection_key` in a table. For a legacy app that does not send App Instance ID, ConnectionSelector generates one when OnServiceStartedCallback of RPC service is received.
+  - When a message is received from ProtocolHandler through OnMessageReceived(), it converts `connection_key` into appropriate App Instance ID and notify ApplicationManager.
+  - It calls OnServiceStartedCallback of ApplicationManager when a new type of service is started on an app. For example, it calls the callback when RPC service is started on a transport for an app, but it will not call when another RPC service is started on a different transport for the same app. Likewise, it calls OnServiceEndedCallback of ApplicationManager when *all* instances of a service for an app are ended.
+* Selection of a connection to send out messages
+  - It receives transmit data from ApplicationManager, looks for the App Instance ID attached with the message and selects the best connection available for the app. Then it converts App Instance ID into connection\_key, create a RawMessage instance and send it to ProtocolHandler. The priority of the connections should be pre-configured, through smartDeviceLink.ini setting.

```

Comment:
Please update to "...connection_key, creates a RawMessage instance and sends it to..."